pages/blog/socketio-automatic-docs.md

```diff
@@ -0,0 +1,553 @@
+---
+title: Flask-SocketIO Automatic Documentation and Validation
+type: Engineering
+tags: ['Python','SocketIO','Flask-SocketIO']
+cover: /img/posts/socketio-automatic-docs/cover.webp
+authors:
+  - name: Daler Rahimov
+    photo: /img/avatars/dalerrahimov.webp
+    link: https://github.com/daler-rahimov/
+    byline: Python Engineer at Keyme LLC
+---
+
+## What is Flask-SocketIO, SocketIO?
+If you are reading this blog you probably already familiar with SocketIO, [python-socketio](https://github.com/miguelgrinberg/python-socketio) and [Flask-SocketIO](https://flask-socketio.readthedocs.io/en/latest/). If you want to learn more about them, please check the following resources:
+ - https://socket.io/docs/v4/
+ - https://www.asyncapi.com/blog/socketio-part1
+ - https://www.asyncapi.com/blog/socketio-part2
+
+## The problem
+Imagine that you are working on a large project that uses a Flask-SocketIO server to handle real-time communication between the client and server. The server was originally well-documented, but over time the documentation has become out of date as the server has evolved and new features have been added.
+
+I found myself in the same situation. I needed to maintain and constantly add more documentation to a Flask-SocketIO server. To make this process more efficient, I sought a solution to automate the generation of documentation from the existing codebase. This would eliminate the need for team members to manually edit the AsyncAPI specification file every time there was a change, a challenging task for those unfamiliar with AsyncAPI. By automating this process, we hoped to save time and reduce the workload for the team.
+
+To address this issue, I decided to implement SIO-AsyncAPI. This tool allows you to generate an AsyncAPI specification from your SocketIO server and validate incoming and outgoing messages against it. This functionality is similar to how FastAPI, Flask-RESTful, and other frameworks have long provided it for RESTful servers. Now, with [SIO-AsyncAPI](https://github.com/daler-rahimov/sio-asyncapi), it is possible to apply this approach to SocketIO servers as well.
+
+## How to use SIO-AsyncAPI
+Instead of giving you a detailed, step-by-step guide, we'll use a more exciting approach. We'll take the existing Flask-SocketIO server and add SIO-AsyncAPI. To make things even more interesting, we'll ask [ChatGPT](https://chat.openai.com/chat) to generate a server for us and use it as a real-world Flask-SocketIO server.
+
+### Generate a Flask-SocketIO server
+I had to ask ChatGPT for multiple alterations in order to get the desired result. Here are my quires to ChatGPT:
```

Comment:
ChatGPT 🤖 strikes again!

Reply:
Yep, I was surprised how good of a code it generated for Flask-SocketIO